### PR TITLE
PACT-598 Switched to a three-part control for start and end dates.

### PR DESCRIPTION
**Pull request by tna-erasmos**

### CODE REVIEW

**Comment by tna-erasmos** [comment]
Whoever merges, could we squash with the following commit message?

```
PACT-598 Switched to a three-part control for start and end dates.
    
Included validation for historical dates which is aware of the switch from Julian to Gregorian calendar in the UK.
```

**Comment by felixgb** [comment]
I think `sbt scalafmtAll` might need to be run on this branch; there are quite a few files that I don't think have been formatted.

**Comment by tna-erasmos** [comment]
You're right, @felixgb ; I'll push another commit.

**Review by felixgb — COMMENTED** [review]
Nice, we discussed in a meeting as well that we should make the covering date parser also parse dates in the Julian / Gregorian calendar, probably in the follow on ticket PACT-621.

File: app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetController.scala [hunk 1]
```diff
@@ -170,4 +177,43 @@ class EditSetController @Inject() (
...
+      case (Some(_), None)                                               => Seq(errorForInvalidEndDate)
+      case (None, Some(_))                                               => Seq(errorForInvalidStartDate)
+      case (None, None) => Seq(errorForInvalidStartDate, errorForInvalidEndDate)
+    }
+    form.copy(errors = form.errors ++ additionErrors)
+  }
+
+  private def message(key: String): String = messagesApi(key)(Lang("en"))
+
+  private def extractStartDate(formValues: Map[String, String]): Option[Date] =
```
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetController.scala:202]: Good idea. Much more "scala".

Will change.

File: app/uk/gov/nationalarchives/omega/editorial/models/EditSet.scala [hunk 2]
```diff
@@ -34,15 +33,22 @@ case class EditSetRecord(
   scopeAndContent: String,
   coveringDates: String,
   formerReferenceDepartment: String,
-  startDate: String,
-  endDate: String
+  startDateDay: String,
```
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/models/EditSet.scala:36]: I'd like to explore this is [the followup ticket](https://national-archives.atlassian.net/browse/PACT-621).

Are there any changes you think should be made now, for this one?

File: test/support/CustomMatchers.scala [hunk 3]
```diff
@@ -85,134 +73,152 @@ object CustomMatchers {
...
+      actualValue = document.select("#coveringDates").attr("value"))
   }
 
   def haveFormerReferenceDepartment(expectedValue: String): Matcher[Document] = (document: Document) => {
-    val actualValue = document.select("#formerReferenceDepartment").attr("value")
+    singleValueMatcher(
+      label = "a former reference department",
+      expectedValue = expectedValue,
+      actualValue = document.select("#formerReferenceDepartment").attr("value"))
+  }
```
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/support/CustomMatchers.scala]: I'll reformat it.

**Review by felixgb — APPROVED** [review]

**Pull request merged**